Prepare curved Poisson and Elasticity systems for use in XCTS equations

nilsvu opened the pull request:
## Proposed changes

The upcoming XCTS equations #2667 make use of the Poisson and Elasticity systems. This PR gets their non-Euclidean terms in order.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Reformulate curved Poisson equation

Move the term that contains Christoffel symbols to the sources.

nilsvu committed:
Add curved terms for elasticity equations

These terms are used in the XCTS equations since the momentum constraint
is an elasticity equation.

nilsvu committed:
Add tag for contraction of Christoffel symbols

nilsdeppe commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -28,11 +28,14 @@ namespace Poisson {
  *
- * where we have chosen the field gradient as an auxiliary variable \f$v_i\f$.
+ * where we have chosen the field gradient as an auxiliary variable \f$v_i\f$
+ * and where \f$\Gamma^i_{jk}=\frac{1}{2}\gamma^{il}\left(\partial_j\gamma_{kl}
```
Just for clarification: you will never be solving for the spatial metric or some function times the spatial metric? Since you have `partial_j gamma_{kl}` my worry is if you need to solve for `f_ij(gamma_{ij})` you then compute the spatial metric from the function you are solving for and then differentiate it. Specifically, if you solve for the conformal factor and `gamma_{ij} = psi^4 \bar{gamma_{ik}}` then this term is no longer a "source" term, I think. Or at least, it's not obvious to me that first derivatives are okay as source terms unless you have an auxiliary variable for `phi_i= partial_i psi` and then compute `partial_i gamma_{kl}` as `4 psi^3 phi_i \bar{gamma}_{kl} + psi^4 \partial_i \bar{gamma}_{kl}`

nilsvu replied:
I'm never solving for the conformal metric `\bar{\gamma}` in the XCTS equations; the conformal metric is fixed. All geometric quantities like Christoffel symbols and the Ricci scalar are defined w.r.t the fixed conformal metric in the XCTS equations. Note that this documentation here is for the curved Poisson equation that has no notion of a "conformal" metric, so it doesn't put a `\bar` on the `\gamma`. In this context the metric `\gamma` is the fixed background of the Poisson equation. When calling this code from the XCTS equations the Poisson background-metric `\gamma` is the XCTS conformal metric `\bar{\gamma}`. Was that sufficiently confusing? :)

nilsdeppe replied:
Okay, to be very clear:
In the XCTS system `Gamma` is the Christoffel symbol associated with the conformal metric?

nilsvu replied:
Yes.

haraldp271 replied:
I am fine with the comment as is.  But seeing Nils D's question, perhaps add a sentence:  The metric gamma_{ij}, and the Christoffel symbols derived from it are assumed to be constants independent of the solution u.

nilsdeppe commented on src/Elliptic/Systems/Elasticity/Equations.hpp:
```diff
@@ -40,9 +40,29 @@ void primal_fluxes(
     const tnsr::I<DataVector, Dim>& coordinates) noexcept;
 
+/*!
+ * \brief Add the contribution \f$-\Gamma^i_{ik}T^{kj} - \Gamma^j_{ik}T^{ik}\f$
```
same concerns about how you compute `Gamma` as for the Poisson system (mostly just a note so if something needs to be changed we remember to change it in both places)

nilsdeppe reviewed (changes requested):
Overall looks fine, just want to make sure we don't accidentally have non-source terms be source terms :)

haraldp271 commented on src/Elliptic/Systems/Elasticity/Equations.cpp:
```diff
@@ -38,6 +37,35 @@ void primal_fluxes(
   }
 }
 
+template <size_t Dim>
```
Line 30, 31 above (cannot comment there):
  d++  ->  ++d,   e++ -> ++e

haraldp271 replied:
Line 28, 29:  suggest to change comment to 
// stress is symmetric whereas flux_for_displacement is not.
// Therefore, must set each component separately.

haraldp271 replied:
Line 32,33 above (cannot comment there):  Stress T^{ij} has upper indices, so don't understand the comment about lower/upper indices.   I would expect the minus-sign already been taken care of by constitutive_relation.stress, so don't understand the comment about the minus sign.  Overall, the entire comment seems superfluous (at least, it doesn't help me understand what's going on).  Suggest to remove entire comment, and place a new comment before Line 27: 
// computes $T^{ij} = - Y^{ijkl} S_{kl}$

haraldp271 commented on src/Elliptic/Systems/Elasticity/Equations.cpp:
```diff
@@ -55,6 +83,21 @@ void auxiliary_fluxes(
+    const tnsr::I<DataVector, Dim>& displacement) noexcept {
+  const auto co_displacement = raise_or_lower_index(displacement, metric);
+  std::fill(flux_for_strain->begin(), flux_for_strain->end(), 0.);
+  for (size_t d = 0; d < Dim; d++) {
```
++d

haraldp271 commented on src/Elliptic/Systems/Elasticity/Equations.cpp:
```diff
@@ -55,6 +83,21 @@ void auxiliary_fluxes(
+  std::fill(flux_for_strain->begin(), flux_for_strain->end(), 0.);
+  for (size_t d = 0; d < Dim; d++) {
+    flux_for_strain->get(d, d, d) += 0.5 * co_displacement.get(d);
+    for (size_t e = 0; e < Dim; e++) {
```
++e

haraldp271 commented on src/Elliptic/Systems/Poisson/Equations.hpp:
```diff
@@ -29,16 +36,28 @@ void euclidean_fluxes(gsl::not_null<tnsr::I<DataVector, Dim>*> flux_for_field,
  * for the curved-space Poisson equation on a spatial metric \f$\gamma_{ij}\f$.
  */
 template <size_t Dim>
 void non_euclidean_fluxes(
```
The terminology 'euclidean' vs. 'non_euclidean' is meant to signal whether Christoffel-symbols and non-trivial metric terms are included.  

However, Christoffelsymbols and non-trivial metrics also arise on **Euclidean** manifolds for **curved** coordinates.  As such, the terminology 'euclidean' vs. 'non_euclidean' isn't mathematically correct.    I don't really have a better suggestion (something like 'kronecker_metric'  or  'euclidean_and_cartesian'  is awful and won't be clearer), and I have no opinion on the level of mathematical rigour spectre is striving for.  @nilsdeppe might want at least to take note.


Update:  I am now unhappy in two regards with the naming euclidean / non_euclidean:
1.  the code-path 'non_euclidean' is also necessary for Euclidean manifolds in curved coordinates (as above)
2. the name 'non_euclidean' says it's *not* euclidean, but does not give information in how it deviates  from 'euclidean'.  The intent is 'curved coordinates'.  However, another easy to make assumption is that non-euclidean refers to **Riemannian**.  


Further, some comments in the code ("\brief Compute the sources \f$S_A\f$ for the Poisson equation on a flat
  metric in Cartesian coordinates.") and even the title of the PR  ("...curved Poisson system...") do not use the euclidean/non_euclidean nomenclature, but rather use more explicit wording. This further indicates that the terminology is not clear.  SpEC uses 'FlatLaplacian' and 'CurvedLaplacian'.

nilsvu replied:
This is precisely the reason I chose `Euclidean` over `Flat`: The connection on a flat space has vanishing coefficients only in Cartesian coordinates. I now realize that the term `Euclidean` doesn't imply Cartesian coordinates but is just a synonym for `Flat`. So my suggestion would be: `FlatCartesian` vs `Curved`. What do you think? I'm happy to hear more suggestions!

nilsdeppe replied:
I can't think of anything else, I like those

haraldp271 replied:
I like FlatCartesian and Curved.   It's nice that the probably more widely used one (Curved) keeps a shorter name.

kidder commented:
@haraldp271 please look at

haraldp271 commented on src/Elliptic/Systems/Poisson/Equations.hpp:
```diff
@@ -16,6 +17,12 @@ class DataVector;
 class er;
 }  // namespace PUP
+namespace Poisson {
+template <size_t Dim, Geometry BackgroundGeometry>
```
Explain the ideas behind the Geometry template parameter.  What is it used for? (activate additional, more costly terms)  What are the requirements on it?  (it must be precisely one of the two tags specified in Elliptic/Systems/Poisson.hpp Geometry).   That kind of meta-information is very helpful to newcomers.   The generic forward declarations here are a good point to place such info, unless it's already elsewhere in a more coherent overview document.

haraldp271 commented on src/Elliptic/Systems/Poisson/Equations.hpp:
```diff
@@ -104,13 +119,14 @@ struct NonEuclideanFluxes {
 /*!
- * \brief Compute the sources \f$S_A\f$ for the Poisson equation.
+ * \brief Compute the sources \f$S_A\f$ for the Poisson equation on a flat
+ * metric in Cartesian coordinates.
```
```suggestion
 * metric in Cartesian coordinates (i.e. Geoemtry::Euclidean).
```

haraldp271 commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -84,18 +87,18 @@ struct FirstOrderSystem {
   // normalize vectors on the faces of an element
+  using inv_metric_tag = tmpl::conditional_t<
+      BackgroundGeometry == Geometry::Euclidean, void,
+      gr::Tags::InverseSpatialMetric<Dim, Frame::Inertial, DataVector>>;
```
These tags are tightly coupled to the choice of Geometry.  Another option would be to turn Geometry::Euclidean and Geometry::NonEuclidean into structs having members 'inv_metric_tag' and 'magnitude_tag'

nilsvu replied:
That's certainly possible, though I think it's cleaner to have the system struct deal with selecting tags and have the enum only provide the cases.

haraldp271 commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -84,18 +87,18 @@ struct FirstOrderSystem {
   // normalize vectors on the faces of an element
+  using inv_metric_tag = tmpl::conditional_t<
+      BackgroundGeometry == Geometry::Euclidean, void,
+      gr::Tags::InverseSpatialMetric<Dim, Frame::Inertial, DataVector>>;
```
Should there be a sanity check to make sure that Geometry has precisely two choices?

nilsvu replied:
I think that would be necessary if the "else"-clause implied a particular enum case, but here the "else" just means: For anything that's not `FlatCartesian`, use the metric. It would keep working even if extra enum cases were added.

nilsvu committed:
Clean up elastic constitutive relation interface

- Return stress by reference, not by value, to avoid allocations
- Remove an unused overload that takes the un-symmetrized
  displacement gradient
- Add an overload that writes the stress in a non-symmetric tensor,
  which is currently how the flux is stored in the DataBox.

nilsvu committed:
Rename Poisson geometry to FlatCartesian/Curved

nilsvu commented:
@haraldp271 @nilsdeppe Please take another look and let me know if your comments were addressed. This is what I did:
- Rebased to fix a conflict.
- Renamed the `Geometry` cases to `FlatCartesian` and `Curved`. Also renamed related functions accordingly.
- Added the commit "Clean up elastic constitutive relation interface" to address some of the comments on code that wasn't touched by the existing commits.
- Added a fixup commit with some documentation clarifications. It also deletes a section that is not quite correct and that will be clarified in upcoming DG-related PRs.

nilsdeppe commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.cpp:
```diff
@@ -37,19 +38,19 @@ tnsr::II<DataVector, 3> IsotropicHomogeneous<3>::stress(
     const tnsr::ii<DataVector, 2>& strain,
     const tnsr::I<DataVector, 2>& /*x*/) const noexcept {
-  auto result = make_with_value<tnsr::II<DataVector, 2>>(strain, 0.);
+  std::fill(stress->begin(), stress->end(), 0.);
```
why zero this? You are assigning to the components below in the for loops.

nilsdeppe commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.cpp:
```diff
@@ -22,13 +22,14 @@ IsotropicHomogeneous<Dim>::IsotropicHomogeneous(double bulk_modulus,
     const tnsr::ii<DataVector, 3>& strain,
     const tnsr::I<DataVector, 3>& /*x*/) const noexcept {
-  auto result = make_with_value<tnsr::II<DataVector, 3>>(strain, 0.);
+  std::fill(stress->begin(), stress->end(), 0.);
```
Why zero this? You are assigning to the components below in the for loops?

nilsdeppe commented on src/PointwiseFunctions/Elasticity/ConstitutiveRelations/CubicCrystal.cpp:
```diff
@@ -27,21 +27,20 @@ CubicCrystal::CubicCrystal(const double c_11, const double c_12,
+void CubicCrystal::stress(const gsl::not_null<tnsr::II<DataVector, 3>*> stress,
+                          const tnsr::ii<DataVector, 3>& strain,
+                          const tnsr::I<DataVector, 3>& /*x*/) const noexcept {
+  std::fill(stress->begin(), stress->end(), 0.);
```
since you're making changes to improve performance, I'd suggest assigning the `(2,2) = (0,0)` outside the loop, have the loop go from 1->3.

nilsdeppe reviewed (commented):
LGTM, if you decided to include any of my suggestions you can squash them in directly

nilsdeppe reviewed (commented):
LGTM!

nilsvu commented:
@haraldp271 Please take a look at the changes and let me know if they addressed your comments (see https://github.com/sxs-collaboration/spectre/pull/2674#issuecomment-762792485 for the list of changes). Then I can rebase, squash the fixup commit and get this ready to merge.

nilsvu commented:
@nilsdeppe @haraldp271 rebased and squashed (after reviewing with @haraldp271 in a call). Please take another look.
